src/caching_key_wrapper.rs

```diff
@@ -160,40 +173,54 @@ where
         None
     }
 
-    async fn cache_encryption_key(&self, bytes: usize, key: DataKey<S>) {
-        let mut cached = self.encryption_cache.lock().await;
+    async fn cache_encryption_key(&self, bytes: usize, key: DataKey<S>, aad: Option<&str>) {
+        let mut encryption_cache = self.encryption_cache.lock().await;
         let mut decryption_cache = self.decryption_cache.lock().await;
 
-        // If the encryption cache has too many entries, remove the first one.
-        // This operation needs to shift all the elements in the Vec, but should
-        // be negligible since the "max_entries" option on the cache will most
-        // likely be <1000.
-        if cached.len() >= self.options.max_entries {
-            cached.remove(0);
-        }
-
-        cached.push(CachedEncryptionEntry {
+        let cache_key = aad.unwrap_or_default().as_bytes();
```

Comment:
If Aad is default, does that mean that all keys will collide in the cache?

Reply:
Keys with same `aad` gets cached together in the same vec:  `lru : { [aad]: Vec<Entries> }`.

Because it's a vec, when a user does concurrent generate keys, we could end up with multiple keys in the vec without colliding, but the next time user does a generate key, we only pop the last entry and serve that, until a limit is reached.

This is the same mechanism as before, just added the separation by `aad`. I hope that clears it?